<a href="https://colab.research.google.com/github/Seeeeiu/Python/blob/master/Lab3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lab 3

In [20]:
import keras
keras.__version__

'2.2.4'

In [0]:
from keras.datasets import mnist

(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

#### Reshape and Scale Features


In [0]:
train_images = train_images.reshape((60000, 28 * 28))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28 * 28))
test_images = test_images.astype('float32') / 255

#### Convert Labels to Dummy Variables

In [0]:
from keras.utils import to_categorical

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

## Example Problem:

#### Specify Model Architecture

In [24]:
from keras import models
from keras import layers

network = models.Sequential()
network.add(layers.Dense(512, activation='relu', input_shape=(28 * 28,)))
network.add(layers.Dense(10, activation='softmax'))
network.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_11 (Dense)             (None, 512)               401920    
_________________________________________________________________
dense_12 (Dense)             (None, 10)                5130      
Total params: 407,050
Trainable params: 407,050
Non-trainable params: 0
_________________________________________________________________


#### Compile Model 

In [0]:
network.compile(optimizer='rmsprop',
                loss='categorical_crossentropy',
                metrics=['accuracy'])

#### Train the Network

In [26]:
import time as t
now=t.time()
network.fit(train_images, train_labels, epochs=5, batch_size=128)
elapsed = t.time()-now
print(elapsed)

Epoch 1/5
60000/60000 [==============================] - 2s 34us/step - loss: 0.2565 - acc: 0.9261
Epoch 2/5
60000/60000 [==============================] - 2s 33us/step - loss: 0.1033 - acc: 0.9693
Epoch 3/5
60000/60000 [==============================] - 2s 33us/step - loss: 0.0682 - acc: 0.9797
Epoch 4/5
60000/60000 [==============================] - 2s 33us/step - loss: 0.0489 - acc: 0.9853
Epoch 5/5
60000/60000 [==============================] - 2s 33us/step - loss: 0.0365 - acc: 0.9889
10.1860511302948


#### Evaluate the Test Data

In [27]:
test_loss, test_acc = network.evaluate(test_images, test_labels)
print('test_acc:', test_acc)

10000/10000 [==============================] - 0s 48us/step
test_acc: 0.9762


## Problem 1:

a. Modify the example model above to contain 3 hidden layers of 512 neurons each.  


In [0]:
network = models.Sequential()
network.add(layers.Dense(512, activation='relu', input_shape=(28 * 28,)))
network.add(layers.Dense(512, activation='relu'))
network.add(layers.Dense(512, activation='relu'))
network.add(layers.Dense(10, activation='softmax'))

network.compile(optimizer='rmsprop',
                loss='categorical_crossentropy',
                metrics=['accuracy'])



b. Specify the time elapsed to train the new model with 5 epochs.

In [29]:
import time as t
now=t.time()
network.fit(train_images, train_labels, epochs=5, batch_size=128)
elapsed = t.time()-now
elapsed

Epoch 1/5
60000/60000 [==============================] - 3s 51us/step - loss: 0.2369 - acc: 0.9261
Epoch 2/5
60000/60000 [==============================] - 3s 47us/step - loss: 0.0902 - acc: 0.9732
Epoch 3/5
60000/60000 [==============================] - 3s 47us/step - loss: 0.0624 - acc: 0.9821
Epoch 4/5
60000/60000 [==============================] - 3s 47us/step - loss: 0.0471 - acc: 0.9863
Epoch 5/5
60000/60000 [==============================] - 3s 47us/step - loss: 0.0392 - acc: 0.9890


14.694356441497803

c. Specify the accuracy on the test data.

In [30]:

test_loss, test_acc = network.evaluate(test_images, test_labels)
print('test_acc:', test_acc)

10000/10000 [==============================] - 1s 55us/step
test_acc: 0.9778


d. Specify the number of parameters in the new model.

In [31]:
network.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_13 (Dense)             (None, 512)               401920    
_________________________________________________________________
dense_14 (Dense)             (None, 512)               262656    
_________________________________________________________________
dense_15 (Dense)             (None, 512)               262656    
_________________________________________________________________
dense_16 (Dense)             (None, 10)                5130      
Total params: 932,362
Trainable params: 932,362
Non-trainable params: 0
_________________________________________________________________


e. If you rerun the fit command, you will begin from the weights you left off with.  Fit the network further by running 5 more epochs.  Then specify the new test accuracy.

In [32]:
network.fit(train_images, train_labels, epochs=5, batch_size=128)

Epoch 1/5
60000/60000 [==============================] - 3s 46us/step - loss: 0.0333 - acc: 0.9910
Epoch 2/5
60000/60000 [==============================] - 3s 47us/step - loss: 0.0297 - acc: 0.9915
Epoch 3/5
60000/60000 [==============================] - 3s 47us/step - loss: 0.0274 - acc: 0.9930
Epoch 4/5
60000/60000 [==============================] - 3s 47us/step - loss: 0.0263 - acc: 0.9935
Epoch 5/5
60000/60000 [==============================] - 3s 47us/step - loss: 0.0246 - acc: 0.9940


In [33]:
test_loss, test_acc = network.evaluate(test_images, test_labels)
print('test_acc:', test_acc)

10000/10000 [==============================] - 0s 50us/step
test_acc: 0.976


f. How does the training accuracy compare with the test set accuracy?

In [34]:
train_loss, train_acc = network.evaluate(train_images, train_labels)
print('train_acc:', train_acc)

60000/60000 [==============================] - 3s 49us/step
train_acc: 0.9939666666666667


We can see that the training accuracy is 0.99397, which is higher than the test set accuracy 0.976.

## Problem 2:

a. Modify the original example problem to use sigmoid nurons in the hidden layer.  How does the accuracy and the (model train) time elapsed compare to the original example?  How do the number of parameters compare?

In [35]:
network = models.Sequential()
network.add(layers.Dense(512, activation='sigmoid', input_shape=(28 * 28,)))
network.add(layers.Dense(10, activation='softmax'))

network.compile(optimizer='rmsprop',
                loss='categorical_crossentropy',
                metrics=['accuracy'])

import time as t
now=t.time()
network.fit(train_images, train_labels, epochs=5, batch_size=128)
elapsed = t.time()-now
print(elapsed)

test_loss, test_acc = network.evaluate(test_images, test_labels)
print('test_acc:', test_acc)
network.summary()

Epoch 1/5
60000/60000 [==============================] - 2s 35us/step - loss: 0.4318 - acc: 0.8820
Epoch 2/5
60000/60000 [==============================] - 2s 33us/step - loss: 0.2348 - acc: 0.9314
Epoch 3/5
60000/60000 [==============================] - 2s 33us/step - loss: 0.1778 - acc: 0.9481
Epoch 4/5
60000/60000 [==============================] - 2s 34us/step - loss: 0.1394 - acc: 0.9589
Epoch 5/5
60000/60000 [==============================] - 2s 33us/step - loss: 0.1129 - acc: 0.9672
10.259517669677734
10000/10000 [==============================] - 1s 51us/step
test_acc: 0.9634
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_17 (Dense)             (None, 512)               401920    
_________________________________________________________________
dense_18 (Dense)             (None, 10)                5130      
Total params: 407,050
Trainable params: 407,050
Non-trainable params: 0
_______

The accuracy using sigmoid activation function is 0.9634, which is lower than the original example accuracy 0.9762. Also the (model train) time elapsed is 10.25, which is higher than original example time elapsed 10.18. The number of parameters of two models are all 407050.

b. Modify the original example problem to use elu nurons in the hidden layer.  How does the accuracy and the (model train) time elapsed compare to the original example?  How do the number of parameters compare?

In [36]:
network = models.Sequential()
network.add(layers.Dense(512, activation='elu', input_shape=(28 * 28,)))
network.add(layers.Dense(10, activation='softmax'))

network.compile(optimizer='rmsprop',
                loss='categorical_crossentropy',
                metrics=['accuracy'])

import time as t
now=t.time()
network.fit(train_images, train_labels, epochs=5, batch_size=128)
elapsed = t.time()-now
print(elapsed)

test_loss, test_acc = network.evaluate(test_images, test_labels)
print('test_acc:', test_acc)
network.summary()


Epoch 1/5
60000/60000 [==============================] - 2s 37us/step - loss: 0.3413 - acc: 0.9006
Epoch 2/5
60000/60000 [==============================] - 2s 34us/step - loss: 0.1773 - acc: 0.9479
Epoch 3/5
60000/60000 [==============================] - 2s 34us/step - loss: 0.1196 - acc: 0.9647
Epoch 4/5
60000/60000 [==============================] - 2s 37us/step - loss: 0.0886 - acc: 0.9728
Epoch 5/5
60000/60000 [==============================] - 2s 37us/step - loss: 0.0714 - acc: 0.9779
10.97841477394104
10000/10000 [==============================] - 1s 59us/step
test_acc: 0.975
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_19 (Dense)             (None, 512)               401920    
_________________________________________________________________
dense_20 (Dense)             (None, 10)                5130      
Total params: 407,050
Trainable params: 407,050
Non-trainable params: 0
_________

The accuracy using elu nurons in the hidden layer is 0.975, which is lower than the original example accuracy 0.9762. Also the (model train) time elapsed is 10.978, which is higher than original example time elapsed 10.18. The number of parameters of two models are all 407050.

c. Modify the original example problem to use the adam optimizer.  How does the accuracy and the (model train) time elapsed compare to the original example?  How do the number of parameters compare?

In [37]:
network = models.Sequential()
network.add(layers.Dense(512, activation='elu', input_shape=(28 * 28,)))
network.add(layers.Dense(10, activation='softmax'))

network.compile(optimizer='adam',
                loss='categorical_crossentropy',
                metrics=['accuracy'])

import time as t
now=t.time()
network.fit(train_images, train_labels, epochs=5, batch_size=128)
elapsed = t.time()-now
print(elapsed)

test_loss, test_acc = network.evaluate(test_images, test_labels)
print('test_acc:', test_acc)
network.summary()

Epoch 1/5
60000/60000 [==============================] - 2s 41us/step - loss: 0.3280 - acc: 0.9052
Epoch 2/5
60000/60000 [==============================] - 2s 37us/step - loss: 0.1760 - acc: 0.9486
Epoch 3/5
60000/60000 [==============================] - 2s 37us/step - loss: 0.1198 - acc: 0.9647
Epoch 4/5
60000/60000 [==============================] - 2s 37us/step - loss: 0.0873 - acc: 0.9737
Epoch 5/5
60000/60000 [==============================] - 2s 36us/step - loss: 0.0696 - acc: 0.9791
11.514878034591675
10000/10000 [==============================] - 1s 54us/step
test_acc: 0.971
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_21 (Dense)             (None, 512)               401920    
_________________________________________________________________
dense_22 (Dense)             (None, 10)                5130      
Total params: 407,050
Trainable params: 407,050
Non-trainable params: 0
________

The accuracy using elu nurons in the hidden layer is 0.971, which is lower than the original example accuracy 0.9762. Also the (model train) time elapsed is 11.51, which is higher than original example time elapsed 10.18. The number of parameters of two models are all 407050.

d. Modify the original example problem to include a second hidden layer with 300 relu neurons, a third hidden layer with 200 neurons and a fourth hidden layer with 100 neurons.  How does the accuracy and the (model train) time elapsed compare to the original example?  How do the number of parameters compare?

In [66]:
network = models.Sequential()
network.add(layers.Dense(512, activation='elu', input_shape=(28 * 28,)))
network.add(layers.Dense(300, activation='relu'))
network.add(layers.Dense(200, activation='relu'))
network.add(layers.Dense(100, activation='relu'))
network.add(layers.Dense(10, activation='softmax'))

network.compile(optimizer='rmsprop',
                loss='categorical_crossentropy',
                metrics=['accuracy'])

import time as t
now=t.time()
network.fit(train_images, train_labels, epochs=5, batch_size=128)
elapsed = t.time()-now
print(elapsed)

test_loss, test_acc = network.evaluate(test_images, test_labels)
print('test_acc:', test_acc)
network.summary()

Epoch 1/5
60000/60000 [==============================] - 3s 51us/step - loss: 0.2595 - acc: 0.9191
Epoch 2/5
60000/60000 [==============================] - 3s 46us/step - loss: 0.1071 - acc: 0.9678
Epoch 3/5
60000/60000 [==============================] - 3s 46us/step - loss: 0.0809 - acc: 0.9757
Epoch 4/5
60000/60000 [==============================] - 3s 46us/step - loss: 0.0641 - acc: 0.9804
Epoch 5/5
60000/60000 [==============================] - 3s 46us/step - loss: 0.0533 - acc: 0.9834
14.518765211105347
10000/10000 [==============================] - 1s 63us/step
test_acc: 0.9774
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_27 (Dense)             (None, 512)               401920    
_________________________________________________________________
dense_28 (Dense)             (None, 300)               153900    
_________________________________________________________________
dense_29 (Den

The accuracy including those hidden layers is 0.9774, which is higher than the original example accuracy 0.9762. Also the (model train) time elapsed is 14.518, which is higher than original example time elapsed 10.18. The number of parameters of this model is 637130, and the original example model is 407050.

## Problem 3:

Now create your own neural network based upon the concepts you learned.  Achieve a higher accuracy than 98% on the test set.

In [0]:
network = models.Sequential()
network.add(layers.Dense(512, activation='elu', input_shape=(28 * 28,)))
network.add(layers.Dense(500, activation='relu'))
network.add(layers.Dense(500, activation='relu'))
network.add(layers.Dense(10, activation='softmax'))

network.compile(keras.optimizers.Adam(lr=0.001, beta_1=0.9, beta_2=0.999, epsilon=None, decay=0.0, amsgrad=False)

,
                loss='binary_crossentropy',
                metrics=['accuracy'])

network.fit(train_images, train_labels, epochs=5, batch_size=128)

test_loss, test_acc = network.evaluate(test_images, test_labels)
print('test_acc:', test_acc)

## Problem 4:

Last year as part of the final exam in Machine Learning I hosted a Kaggle competition on the Plankton challenge: https://www.kaggle.com/t/830c8dd97ae748f8bd3fd52e8ccca849

a. **Using only the first 1600 features**, use the keras methods and models that have been used thus far make a submission to the competition.  Report your accuracy (your score will not show on the leaderboard since this competition is closed).

In [73]:
import pandas as pd
import io

from google.colab import drive
drive.mount('/content/gdrive')

train = pd.read_csv('/content/gdrive/My Drive/lab3/train.csv', sep=',', decimal='.')
test = pd.read_csv('/content/gdrive/My Drive/lab3/test.csv', sep=',', decimal='.')
train_class = pd.get_dummies(train.iloc[:,0])


train = train.iloc[:,1:1601]
test = test.iloc[:,1:1601]



Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [68]:
from keras import models
from keras import layers

network = models.Sequential()
network.add(layers.Dense(512, activation='relu', input_shape=(1600,)))
network.add(layers.Dense(512, activation='relu'))
network.add(layers.Dense(512, activation='relu'))
network.add(layers.Dense(121, activation='softmax'))

network.compile(optimizer='adam',
                loss='categorical_crossentropy',
                metrics=['accuracy'])

network.fit(train, train_class, epochs=10, batch_size=128)
test_class = network.predict(test)


Epoch 1/10
16789/16789 [==============================] - 2s 145us/step - loss: 3.7016 - acc: 0.1534
Epoch 2/10
16789/16789 [==============================] - 2s 111us/step - loss: 2.7299 - acc: 0.2956
Epoch 3/10
16789/16789 [==============================] - 2s 112us/step - loss: 2.3151 - acc: 0.3686
Epoch 4/10
16789/16789 [==============================] - 2s 111us/step - loss: 2.0152 - acc: 0.4310
Epoch 5/10
16789/16789 [==============================] - 2s 111us/step - loss: 1.7479 - acc: 0.4949
Epoch 6/10
16789/16789 [==============================] - 2s 112us/step - loss: 1.5064 - acc: 0.5566
Epoch 7/10
16789/16789 [==============================] - 2s 110us/step - loss: 1.2890 - acc: 0.6157
Epoch 8/10
16789/16789 [==============================] - 2s 111us/step - loss: 1.0813 - acc: 0.6740
Epoch 9/10
16789/16789 [==============================] - 2s 111us/step - loss: 0.8794 - acc: 0.7313
Epoch 10/10
16789/16789 [==============================] - 2s 110us/step - loss: 0.7006 - a

In [0]:
import numpy as np
pred = np.argmax(test_class,axis=1)

In [0]:
submission = pd.read_csv('/content/gdrive/My Drive/lab3/sample_submission.csv', sep=',', decimal='.')
submission['Class']=pred+1
df = pd.DataFrame(data=submission)
df.to_csv('/content/gdrive/My Drive/lab3/out.csv',index = False)

My accuray: 0.37172

b. Try another method that you think would be interesting (you can do whatever model/features you want for this problem but it must be different than the previous problem).  

In [74]:
# another neuron network
train = pd.read_csv('/content/gdrive/My Drive/lab3/train.csv', sep=',', decimal='.')
test = pd.read_csv('/content/gdrive/My Drive/lab3/test.csv', sep=',', decimal='.')
train_class = pd.get_dummies(train.iloc[:,0])


train = train.iloc[:,1:1601]
test = test.iloc[:,1:1601]

from keras import models
from keras import layers

network = models.Sequential()
network.add(layers.Dense(1024, activation='relu', input_shape=(1600,)))
network.add(layers.Dense(1024, activation='relu'))
network.add(layers.Dense(121, activation='softmax'))

network.compile(optimizer='nadam',
                loss='categorical_crossentropy',
                metrics=['accuracy'])

network.fit(train, train_class, epochs=15, batch_size=128)
test_class = network.predict(test)


Epoch 1/15
16789/16789 [==============================] - 3s 172us/step - loss: 3.4152 - acc: 0.1972
Epoch 2/15
16789/16789 [==============================] - 2s 141us/step - loss: 2.4616 - acc: 0.3501
Epoch 3/15
16789/16789 [==============================] - 2s 141us/step - loss: 1.9717 - acc: 0.4493
Epoch 4/15
16789/16789 [==============================] - 2s 141us/step - loss: 1.5544 - acc: 0.5484
Epoch 5/15
16789/16789 [==============================] - 2s 141us/step - loss: 1.1691 - acc: 0.6471
Epoch 6/15
16789/16789 [==============================] - 2s 140us/step - loss: 0.8279 - acc: 0.7470
Epoch 7/15
16789/16789 [==============================] - 2s 145us/step - loss: 0.5783 - acc: 0.8242
Epoch 8/15
16789/16789 [==============================] - 3s 154us/step - loss: 0.4103 - acc: 0.8759
Epoch 9/15
16789/16789 [==============================] - 3s 152us/step - loss: 0.3038 - acc: 0.9100
Epoch 10/15
16789/16789 [==============================] - 3s 150us/step - loss: 0.2137 - a

In [0]:
df = pd.DataFrame(test_class)
test_class = df.idxmax(axis=1)

In [0]:
import numpy as np
submission = pd.read_csv('/content/gdrive/My Drive/lab3/sample_submission.csv', sep=',', decimal='.')
submission['Class']=test_class+1
df = pd.DataFrame(data=submission)
df.to_csv('/content/gdrive/My Drive/lab3/out2.csv',index = False)

In [0]:
Accuray: 0.35345

## Problem 5:

For this problem we will be working with model churn data. Import the libraries and load in the model churn data by running the next 2 following code chunks.

In [0]:
# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [110]:
import pandas as pd
import io

from google.colab import drive
drive.mount('/content/gdrive')

dataset = pd.read_csv('/content/gdrive/My Drive/Colab Notebooks/Churn_Modelling.csv', sep=',', decimal='.')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


a. Show the first 10 observations of the original data set (dataset).

In [111]:
dataset.iloc[0:10,:]

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   
5          6    15574012       Chu          645     Spain    Male   44   
6          7    15592531  Bartlett          822    France    Male   50   
7          8    15656148    Obinna          376   Germany  Female   29   
8          9    15792365        He          501    France    Male   44   
9         10    15592389        H?          684    France    Male   27   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   
5       8  113755.78              2          1               0   
6       7       0.00              2          1               1   
7       4  115046.74              4          1               0   
8       4  142051.07              2          0               1   
9       2  134603.88              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0  
5        149756.71       1  
6         10062.80       0  
7        119346.88       1  
8         74940.50       0  
9         71725.73       0

b. Which 3 columns should not be used as features? Create the appropriate X matrix (including categorical variables) and the y vector as the last column.

In [0]:
#RowNumber CustomerId Surname
X = dataset.iloc[:,3:13].values
y = dataset.iloc[:,13].values

c. Print out the first 10 observations (and the first 6 columns) of the X matrix and comment on the number of observations and number of columns.

In [113]:
print(X[:10,0:6])
print(X.shape)

[[619 'France' 'Female' 42 2 0.0]
 [608 'Spain' 'Female' 41 1 83807.86]
 [502 'France' 'Female' 42 8 159660.8]
 [699 'France' 'Female' 39 1 0.0]
 [850 'Spain' 'Female' 43 2 125510.82]
 [645 'Spain' 'Male' 44 8 113755.78]
 [822 'France' 'Male' 50 7 0.0]
 [376 'Germany' 'Female' 29 4 115046.74]
 [501 'France' 'Male' 44 4 142051.07]
 [684 'France' 'Male' 27 2 134603.88]]
(10000, 10)


d. After running the following code in the next code chunk, print out the first 10 observations (and the first 6 columns).  Also, print out the shape of the X matrix.

In [0]:
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelEncoderX1=LabelEncoder()
X[:,1]=labelEncoderX1.fit_transform(X[:,1])

labelEncoderX1=LabelEncoder()
X[:,2]=labelEncoderX1.fit_transform(X[:,2])

In [115]:
print(X[:10,0:6])
print(X.shape)

[[619 0 0 42 2 0.0]
 [608 2 0 41 1 83807.86]
 [502 0 0 42 8 159660.8]
 [699 0 0 39 1 0.0]
 [850 2 0 43 2 125510.82]
 [645 2 1 44 8 113755.78]
 [822 0 1 50 7 0.0]
 [376 1 0 29 4 115046.74]
 [501 0 1 44 4 142051.07]
 [684 0 1 27 2 134603.88]]
(10000, 10)


e. How do the X matrices from c and d compare? What does the one in column 2 denote?

f. Run the following code and again print out the first 10 observations and 6 columns. Also print the shape of the X matrix.

In [116]:
onehotencoder = OneHotEncoder(categorical_features = [1])
X = onehotencoder.fit_transform(X).toarray()

/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/_encoders.py:368: FutureWarning: The handling of integer data will change in version 0.22. Currently, the categories are determined based on the range [0, max(values)], while in the future they will be determined based on the unique values.
If you want the future behaviour and silence this warning, you can specify "categories='auto'".
In case you used a LabelEncoder before this OneHotEncoder to convert the categories to integers, then you can now use the OneHotEncoder directly.
  warnings.warn(msg, FutureWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/_encoders.py:390: DeprecationWarning: The 'categorical_features' keyword is deprecated in version 0.20 and will be removed in 0.22. You can use the ColumnTransformer instead.
  "use the ColumnTransformer instead.", DeprecationWarning)


In [117]:
print(X[:10,0:6])
print(X.shape)

[[  1.   0.   0. 619.   0.  42.]
 [  0.   0.   1. 608.   0.  41.]
 [  1.   0.   0. 502.   0.  42.]
 [  1.   0.   0. 699.   0.  39.]
 [  0.   0.   1. 850.   0.  43.]
 [  0.   0.   1. 645.   1.  44.]
 [  1.   0.   0. 822.   1.  50.]
 [  0.   1.   0. 376.   0.  29.]
 [  1.   0.   0. 501.   1.  44.]
 [  1.   0.   0. 684.   1.  27.]]
(10000, 12)


g. Explain why we should get rid of the first column.  Get rid of the first column in the X matrix.  Print out the first 6 observations after getting rid of the first column.

In [118]:
X = X[:,1:]
print(X[:6,:])

[[0.0000000e+00 0.0000000e+00 6.1900000e+02 0.0000000e+00 4.2000000e+01
  2.0000000e+00 0.0000000e+00 1.0000000e+00 1.0000000e+00 1.0000000e+00
  1.0134888e+05]
 [0.0000000e+00 1.0000000e+00 6.0800000e+02 0.0000000e+00 4.1000000e+01
  1.0000000e+00 8.3807860e+04 1.0000000e+00 0.0000000e+00 1.0000000e+00
  1.1254258e+05]
 [0.0000000e+00 0.0000000e+00 5.0200000e+02 0.0000000e+00 4.2000000e+01
  8.0000000e+00 1.5966080e+05 3.0000000e+00 1.0000000e+00 0.0000000e+00
  1.1393157e+05]
 [0.0000000e+00 0.0000000e+00 6.9900000e+02 0.0000000e+00 3.9000000e+01
  1.0000000e+00 0.0000000e+00 2.0000000e+00 0.0000000e+00 0.0000000e+00
  9.3826630e+04]
 [0.0000000e+00 1.0000000e+00 8.5000000e+02 0.0000000e+00 4.3000000e+01
  2.0000000e+00 1.2551082e+05 1.0000000e+00 1.0000000e+00 1.0000000e+00
  7.9084100e+04]
 [0.0000000e+00 1.0000000e+00 6.4500000e+02 1.0000000e+00 4.4000000e+01
  8.0000000e+00 1.1375578e+05 2.0000000e+00 1.0000000e+00 0.0000000e+00
  1.4975671e+05]]


h. Perform a train_test_split with test set of 20% of the data.

In [0]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = .2, random_state = 0 ) # random_state:setting the seed 

i. Below is code to scale the features in both the training and testing set using the StandardScaler. What is the difference between fitting the transform (fit_transform) and applying the transform (transform)?

In [0]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train_sc = sc.fit_transform(X_train)
X_test_sc = sc.transform(X_test)

j. Use keras to train a model and report the test set accuracy.

In [126]:
from keras import models
from keras import layers
y_train=pd.get_dummies(y_train)
y_test =pd.get_dummies(y_test)
network = models.Sequential()
network.add(layers.Dense(8000, activation='relu',input_shape = (11,)))
network.add(layers.Dense(512, activation='relu'))
network.add(layers.Dense(2, activation='sigmoid'))
network.compile(optimizer='adam',
                loss='binary_crossentropy',
                metrics=['accuracy'])
network.fit(X_train_sc, y_train, epochs=7,batch_size = 10)

test_loss, test_acc = network.evaluate(X_test_sc, y_test)
print('test_acc:', test_acc)

Epoch 1/7
8000/8000 [==============================] - 8s 1ms/step - loss: 0.3976 - acc: 0.8339
Epoch 2/7
8000/8000 [==============================] - 7s 916us/step - loss: 0.3583 - acc: 0.8540
Epoch 3/7
8000/8000 [==============================] - 7s 916us/step - loss: 0.3521 - acc: 0.8574
Epoch 4/7
8000/8000 [==============================] - 7s 918us/step - loss: 0.3456 - acc: 0.8581
Epoch 5/7
8000/8000 [==============================] - 7s 919us/step - loss: 0.3427 - acc: 0.8632
Epoch 6/7
8000/8000 [==============================] - 7s 915us/step - loss: 0.3386 - acc: 0.8632
Epoch 7/7
2000/2000 [==============================] - 0s 180us/step
test_acc: 0.8615


## Problem 6:

For the breast cancer dataset (available at https://archive.ics.uci.edu/ml/datasets/Breast+Cancer+Wisconsin+%28Diagnostic%29) do the following:
* Use column 1 as the response variable and all other values (except column 0) as features
* Scale the feature variables
* Split the data with 75% of the training data in the training set and 25% in the test set
* Apply PCA to the training set so that 90% of the variance is retained
* Apply neural networks to the training PCA dataset to predict the response variable
* Predict the test data set and print out the accuracy

In [84]:
import pandas as pd
import io

from google.colab import drive
drive.mount('/content/gdrive')

data = pd.read_csv('/content/gdrive/My Drive/lab3/breastcancer.csv', sep=',', decimal='.', header=None)

data.replace(('M', 'B'), (1, 0), inplace=True)

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [85]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler() 
df = scaler.fit_transform(data) 

/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/data.py:625: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
/usr/local/lib/python3.6/dist-packages/sklearn/base.py:462: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)


In [0]:
y = data.values[:,1]
data = data.values[:,2:32]


In [0]:
import pandas as pd
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(data, y, test_size=0.25)

In [0]:
from sklearn.decomposition import PCA
pca = PCA(0.90)

X_train_pca = pca.fit_transform(X_train)
X_test_pca = pca.transform(X_test)


In [104]:
from keras import models
from keras import layers
y_train=pd.get_dummies(y_train)
y_test =pd.get_dummies(y_test)
network = models.Sequential()
network.add(layers.Dense(700, activation='relu',input_shape = (1,)))
network.add(layers.Dense(512, activation='relu'))
network.add(layers.Dense(2, activation='sigmoid'))
network.compile(optimizer='adam',
                loss='binary_crossentropy',
                metrics=['accuracy'])
network.fit(X_train_pca, y_train, epochs=7,batch_size = 10)

Epoch 1/7
426/426 [==============================] - 1s 2ms/step - loss: 1.3949 - acc: 0.8768
Epoch 2/7
426/426 [==============================] - 0s 479us/step - loss: 1.3487 - acc: 0.9061
Epoch 3/7
426/426 [==============================] - 0s 482us/step - loss: 1.2338 - acc: 0.8991
Epoch 4/7
426/426 [==============================] - 0s 456us/step - loss: 1.2672 - acc: 0.8991
Epoch 5/7
426/426 [==============================] - 0s 554us/step - loss: 1.3391 - acc: 0.9026
Epoch 6/7
426/426 [==============================] - 0s 461us/step - loss: 1.1926 - acc: 0.9038
Epoch 7/7
426/426 [==============================] - 0s 451us/step - loss: 1.1292 - acc: 0.8979


In [106]:
test_loss, test_acc = network.evaluate(X_test_pca, y_test)
print('test_acc:', test_acc)

143/143 [==============================] - 0s 1ms/step
test_acc: 0.9090909090909091
